src/style/dimension_ops.rs

```diff
@@ -0,0 +1,329 @@
+//! Implementations for the standard operations `+`, `-`, `*` and `/` for [`Dimension`].
+
+use super::CalcDimension;
+use super::Dimension;
+use core::ops::{Add, Div, Mul, Sub};
+
+impl Add<Dimension> for Dimension {
+    type Output = Dimension;
+
+    /// Add two [`Dimension`]s together.
+    ///
+    /// If the result of the calculation can already be known, it is calculated immediately.
+    /// Otherwise, variants of [`Dimension::Calc`] are constructed.
+    ///
+    /// # Examples
+    ///
+    /// When the result can already be known, it is calculated immediately:
+    ///
+    /// ```
+    /// # use taffy::style::Dimension;
+    /// #
+    /// let result = Dimension::Points(13.0) + Dimension::Points(5.5);
+    /// assert_eq!(result, Dimension::Points(18.5));
+    /// ```
+    ///
+    /// If any of the operants is [`Dimension::Undefined`], the result will be [`Dimension::Undefined`]:
+    ///
+    /// ```
+    /// # use taffy::style::Dimension;
+    /// #
+    /// let result = Dimension::Points(28.0) + Dimension::Undefined;
+    /// assert_eq!(result, Dimension::Undefined);
+    /// ```
+    ///
+    /// The behavior is analogue for [`Dimension::Auto`].
+    ///
+    /// If the result cannot be known beforehand, variants of [`Dimension::Calc`] will be constructed:
+    ///
+    /// ```
+    /// # use taffy::style::Dimension;
+    /// # use taffy::style::CalcDimension;
+    /// #
+    /// let result = Dimension::Points(5.0) + Dimension::Percent(80.0);
+    /// let expected = Dimension::Calc(
+    ///     CalcDimension::Add(
+    ///         Box::new(Dimension::Points(5.0)),
+    ///         Box::new(Dimension::Percent(80.0))
+    ///     )
+    /// );
+    /// assert_eq!(result, expected);
+    /// ```
+    fn add(self, rhs: Dimension) -> Self::Output {
+        match (self, rhs) {
+            (Dimension::Undefined, _) | (_, Dimension::Undefined) => Dimension::Undefined,
+            (Dimension::Auto, _) | (_, Dimension::Auto) => Dimension::Auto,
+            (Dimension::Points(lhs), Dimension::Points(rhs)) => Dimension::Points(lhs + rhs),
+            (Dimension::Percent(lhs), Dimension::Percent(rhs)) => Dimension::Percent(lhs + rhs),
+            // Result can't be known in advance, defer calculation until resolving the actual values
+            (lhs, rhs) => Dimension::Calc(CalcDimension::Add(Box::new(lhs), Box::new(rhs))),
```

Comment:
I don't think evaluating a calc expression should allocate new boxes. It should be possible to do this without allocating with a recursive algorithm.

Reply:
(same for other operations (sub, mul, div)

Reply:
I don't think this is possible, because `CalcDimension` is a recursive type. We need the `Box`es to have the size known at compile-time.

At least I haven't found a way around this problem yet.

Reply:
Good idea, done!